GG/src/ScrollPanel.cpp

```diff
@@ -0,0 +1,116 @@
+#include "../GG/ScrollPanel.h"
+
+#include <GG/DrawUtil.h>
+#include <GG/Flags.h>
+#include <GG/Scroll.h>
+#include <GG/Clr.h>
+#include <GG/ClrConstants.h>
+#include <GG/WndEvent.h>
+#include <GG/StyleFactory.h>
+
+#include <algorithm>
+
+namespace GG {
+
+    using std::max;
+    using std::min;
+
+    namespace {
+        // Minimum acceptable width for a scroll.
+        X MIN_SCROLL_WIDTH(14);
+
+        // Space to leave left of the vertical scroll.
+        X SCROLL_MARGIN_X(2);
+    }
+
+    void ScrollPanel::SizeMove(const Pt& ul, const Pt& lr)
+    {
+        // Store the new size for input to the layout.
+        Wnd::SizeMove(ul, lr);
+
+        // Lay out the children.
+        DoLayout();
+    }
+
+    void ScrollPanel::Render() {
+        FlatRectangle(UpperLeft(), LowerRight(), m_background_color, CLR_ZERO, 0);
+    }
+
+    ScrollPanel::ScrollPanel() : m_vscroll(0), m_content(0), m_background_color(CLR_ZERO) {}
+
+    ScrollPanel::ScrollPanel(X x, Y y, X w, Y h, Wnd* content)
+    : Wnd(x, y, w, h, INTERACTIVE), m_vscroll(0), m_content(content), m_background_color(CLR_ZERO)
+    {
+        // Very important to clip the content of this panel,
+        // to actually only show the currently viewed part.
+        SetChildClippingMode(ClipToClient);
+
+        // Get the scroll bar from the current style factory.
+        boost::shared_ptr<StyleFactory> style = GetStyleFactory();
+        m_vscroll = style->NewMultiEditVScroll(CLR_WHITE, CLR_BLACK);
+
+        // Don't accept less than MIN_SCROLL_WIDTH pixels wide scrolls.
+        if (m_vscroll->Width() < MIN_SCROLL_WIDTH) {
+            m_vscroll->Resize(GG::Pt(MIN_SCROLL_WIDTH, m_vscroll->Height()));
+        }
+
+        AttachChild(m_vscroll);
+        AttachChild(content);
+
+        Connect(m_vscroll->ScrolledSignal, &ScrollPanel::OnScrolled, this);
+
+        DoLayout();
+    }
+
+    void GG::ScrollPanel::ScrollTo(Y pos)
+    {
+        m_vscroll->ScrollTo(Value(pos));
+        SignalScroll(*m_vscroll, true);
+    }
+
+    void ScrollPanel::SetBackgroundColor(const Clr& color) { m_background_color = color; }
```

Comment:
be consistent about how single-line function implementations are formatted. See ScrollPanel::Render above.